include/zephyr/sys/zync.h

```diff
@@ -0,0 +1,386 @@
+/* Copyright (c) 2022 Google LLC.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+#ifndef ZEPHYR_SYS_ZYNC_H
+#define ZEPHYR_SYS_ZYNC_H
+
+#include <string.h>
+#include <zephyr/sys/util_macro.h>
+#include <zephyr/sys/atomic.h>
+#include <zephyr/spinlock.h>
+#include <zephyr/sys_clock.h>
+#include <zephyr/kernel_structs.h>
+#include <zephyr/syscall.h>
+#include <errno.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define K_ZYNC_ATOM_VAL_BITS 24
+#define K_ZYNC_ATOM_VAL_MAX ((int32_t)(BIT(K_ZYNC_ATOM_VAL_BITS) - 1))
+
+/** @brief Zephyr atomic synchronization primitive
+ *
+ * The zync atom stores the counted state variable for a struct
+ * k_zync, in such a way that it can be atomically modified.
+ */
+typedef union {
+	atomic_t atomic;
+	struct {
+		uint32_t val : K_ZYNC_ATOM_VAL_BITS; /* Value of the lock/counter */
+		bool waiters : 1;                    /* Threads waiting? */
+	};
+} k_zync_atom_t;
+
+/* True if zyncs must track their "owner" */
+#if defined(CONFIG_ZYNC_PRIO_BOOST) || defined(CONFIG_ZYNC_RECURSIVE) \
+	|| defined(CONFIG_ZYNC_VALIDATE)
+#define Z_ZYNC_OWNER 1
+#endif
+
+/* True if all zync calls must go through the full kernel call
+ * (i.e. the atomic shortcut can't be used)
+ */
+#if defined(CONFIG_ZYNC_RECURSIVE) || defined(CONFIG_ZYNC_MAX_VAL) \
+	|| defined(CONFIG_ZYNC_PRIO_BOOST) \
+	|| (defined(CONFIG_ZYNC_USERSPACE_COMPAT) && defined(CONFIG_USERSPACE))
+#define Z_ZYNC_ALWAYS_KERNEL 1
+#endif
+
+/* True if every k_zync struct includes its own atom (it's not in the
+ * zync_pair to make all the vrfy boilerplate simpler)
+ */
+#if defined(Z_ZYNC_ALWAYS_KERNEL) || !defined(CONFIG_USERSPACE)
+#define Z_ZYNC_INTERNAL_ATOM 1
+#endif
+
+struct k_zync_cfg {
+	uint32_t atom_init : K_ZYNC_ATOM_VAL_BITS;
+	bool fair : 1;
+	IF_ENABLED(CONFIG_ZYNC_PRIO_BOOST, (bool prio_boost : 1;))
+	IF_ENABLED(CONFIG_ZYNC_RECURSIVE, (bool recursive : 1;))
+	IF_ENABLED(CONFIG_ZYNC_MAX_VAL, (uint32_t max_val;))
+};
+
+/* @brief Fundamental Zephyr thread synchronization primitive
+ *
+ * @see `k_zync()`
+ */
+struct k_zync {
+	_wait_q_t waiters;
+	IF_ENABLED(Z_ZYNC_OWNER, (struct k_thread *owner;))
+	IF_ENABLED(CONFIG_POLL, (sys_dlist_t poll_events;))
+	struct k_spinlock lock;
+	struct k_zync_cfg cfg;
+	IF_ENABLED(CONFIG_ZYNC_RECURSIVE, (uint32_t rec_count;))
+	IF_ENABLED(Z_ZYNC_INTERNAL_ATOM, (k_zync_atom_t atom;))
+	IF_ENABLED(CONFIG_ZYNC_PRIO_BOOST, (int8_t orig_prio;))
+	IF_ENABLED(CONFIG_POLL, (bool pollable;))
+};
+
+#define Z_ZYNC_MVCLAMP(v) ((v) == 0 ? K_ZYNC_ATOM_VAL_MAX \
+			      : CLAMP((v), 0, K_ZYNC_ATOM_VAL_MAX))
+
+#define K_ZYNC_INITIALIZER(init, isfair, rec, prioboost, maxval) {	\
+	.cfg.atom_init = init,						\
```

Comment:
If I am understanding the coding standards properly (Rule 20.7), init should be (init). I only point this out because of certification efforts. There are are few other similar cases speckled throughout.

Reply:
Yup, fixed.  Looked for and found a few others, but please call them out if you see more.